test/distributed/org/apache/cassandra/distributed/test/sai/PartialUpdateHandlingTest.java

```diff
@@ -0,0 +1,401 @@
+            return node;
+        }
+
+        private int deleteReplica(int node, Object column)
+        {
+            currentCells.remove((String) column);
+            String dml = String.format("DELETE %s FROM %s.%s WHERE pk = %d and ck = 0", 
+                                       column, KEYSPACE, specification.tableName(), specification.partitionKey);
+
+            if (column.equals("s"))
+                dml = String.format("DELETE %s FROM %s.%s WHERE pk = %d", 
+                                    column, KEYSPACE, specification.tableName(), specification.partitionKey);
+
+            CLUSTER.get(node++).executeInternal(dml);
+
+            if (specification.readCL == QUORUM)
+                CLUSTER.get(node).executeInternal(dml);
+            return node;
+        }
+
+        public void validateCurrent()
+        {
+            Object[][] result = validateModel(currentCells);
+
+            // Hard-code the partition key and clustering key:
+            List<Object> expectedRow = Lists.newArrayList(specification.partitionKey, 0);
+
+            for (Object column : specification.nonKeyColumns())
+                expectedRow.add(currentCells.get((String) column));
+
+            assertRows(result, expectedRow.toArray());
+        }
+        
+        public void validatePrevious()
+        {
+            // Ensure queries against the previous version of the row no longer match.
+            assertRows(validateModel(previousCells));
+        }
+
+        private Object[][] validateModel(Map<String, Integer> cells)
+        {
+            StringBuilder select = new StringBuilder("SELECT pk, ck");
+
+            for (Object column : specification.nonKeyColumns())
+                select.append(", ").append(column);
+
+            select.append(" FROM ").append(KEYSPACE).append('.').append(specification.tableName()).append(" WHERE ");
+
+            ArrayList<String> restricted = Lists.newArrayList(specification.columns);
+            if (specification.restrictPartitionKey)
+                restricted.add("pk");
+
+            List<String> clauses = restricted.stream().map(c -> c + " = " + cells.get(c)).collect(Collectors.toList());
+            select.append(String.join(" AND ", clauses));
+
+            if (specification.readCL == QUORUM)
+            {
+                // Make sure node 2, which is the only complete replica, can't participate in a query...
+                CLUSTER.get(2).runOnInstance(() -> {
+                    Keyspace keyspace = Objects.requireNonNull(Schema.instance.getKeyspaceInstance(KEYSPACE));
+                    SecondaryIndexManager sim = keyspace.getColumnFamilyStore(TEST_TABLE_NAME).indexManager;
+                    for (String s : Arrays.asList("ck_idx", "s_idx", "a_idx", "b_idx"))
+                        sim.makeIndexNonQueryable(sim.getIndexByName(s), Index.Status.BUILD_FAILED);
+                });
+
+                SAIUtil.waitForIndexNonQueryable(CLUSTER, KEYSPACE, 2);
```

Comment:
There might be an even better way, but this avoids shutdown/startup overhead and makes it much faster to get through the entire suite.